app/services/chf/create_dzi_service.rb

```diff
@@ -0,0 +1,263 @@
+  #  * dzi_job_tmp_dir
+  #  * dzi_s3_bucket
+  #  * dzi_s3_bucket_region
+  #
+  # Seek rake chf:dzi for tools for managing the S3 bucket contents.
+  #
+  # MAYBE would we better off using actual libvips bindings at https://github.com/jcupitt/ruby-vips
+  #   instead of shell out? I think it probably doesn't matter.
+  #
+  # TODO what do we want to do in dev? create dzi's to public/? No DZIs?
+  #
+  # TODO some cleverer concurrency stuff if two of these jobs try acting at the same
+  # time, keep the out of using each others files, or let them actually share/wait
+  # on each other files.
+  class CreateDziService
+    WORKING_DIR = CHF::Env.lookup(:dzi_job_tmp_dir)
+    UPLOAD_THREADS = 128
+
+    class_attribute :vips_command
+    self.vips_command = "vips"
+
+    class_attribute :jpeg_quality
+    self.jpeg_quality = "85"
+
+    class_attribute :suppress_vips_stderr
+    self.suppress_vips_stderr = true
+
+    class_attribute :cache_control
+    self.cache_control = "max-age=31536000"
+
+    class_attribute :acl
+    self.acl = 'public-read'
+
+
+    attr_accessor :file_id, :checksum, :s3_bucket
+
+
+    def initialize(file_id, checksum: nil)
+      raise ArgumentError("file_id arg can not be empty") if file_id.blank?
+      @file_id = file_id
+      @checksum = checksum
+      @s3_bucket = self.class.s3_bucket!
+    end
+
+    # Downloads datastream and creates DZI files in WORKING_DIR,
+    # uploads them to S3.
+    #
+    # If lazy is true, will first check to see if the .dzi file already exists
+    # on S3, and if it does skip generation and upload. It doesn't check
+    # to make sure all tiles are correct, just dzi file exists.
+    def call(lazy: false)
+      if lazy && s3_bucket.object(dzi_file_name).exists?
+        return false
+      end
+
+      ensure_dirs
+
+      fetch_from_fedora!
+
+      create_dzi!
+
+      upload_to_s3!
+
+      return true
+    ensure
+      clean_up_tmp_files
+    end
+
+    protected
+
+    # Fetch Hydra::PCDM::File from fedora and save to WORKING_DIR,
+    # streaming for efficiency.
+    def fetch_from_fedora!
+      response = nil
+      # We hope this is the right way to get the actual uri to fetch cheaply?
+      uri = URI.parse(FileSet.translate_id_to_uri.call(file_id))
+
+      fedora_fetch_benchmark = Benchmark.measure do
+        response = Net::HTTP.start(uri.host, uri.port) do |http|
+          request = Net::HTTP::Get.new uri
+
+          if ActiveFedora.fedora.user || ActiveFedora.fedora.password
+            request.basic_auth(ActiveFedora.fedora.user, ActiveFedora.fedora.password)
+          end
+
+          http.request request do |response|
+            open local_original_file_path, 'wb' do |io|
+              response.read_body do |chunk|
+                io.write chunk
+              end
+            end
+          end
+        end
+      end
+      Rails.logger.debug("#{self.class.name}: fetch_from_fedora: #{fedora_fetch_benchmark}")
+      unless response.code == "200"
```

Comment:
not sure, but good call to be more defensive, changed to `unless response && response.code == "200"`